### ISSUE TITLE
automatic overlay editing from a container exec or container shell

### ISSUE THREAD

**Issue by anderbubble** [hpcng/warewulf#735]
> It would [...] be nice if you made overlays something you could engage with a little more interactively and in context. e.g., use an "overlay" mount (https://www.kernel.org/doc/html/latest/filesystems/overlayfs.html) with the "container" folder structure as the read-only "lower" layer and a specified overlay as the read-write "upper" via a command like:
>
> `wwctl container exec <my container> --overlay <my overlay> /bin/bash`
>
> Then you're in an interactive shell, the full context of the merged filesystem is available, you can use system tools inside either the container or the overlay to modify the contents of the overlay, and when you exit, the only thing that needs to be rebuilt is the overlay since it was the only writeable thing anyhow. I think this would allow for more interesting applications of overlays, especially if one could apply more than one "system"-type overlay to a node with an order of precedence.

From @elfurbe

**Comment by elfurbe** [comment]
I think github ate some formatting, should theoretically look like:
`wwctl container exec <my container> --overlay <my overlay> /bin/bash`

**Comment by anderbubble** [comment]
Fixed.

**Comment by mslacken** [comment]
We have a kind of this command in the actual development branch with 
```
wwctl container exec --node <my-node> <mycontainer> /bin/bash
```
which will give you a read only shell in the container with the rendered overlay for the given node. 
It has to be read only, as if you would change a rendered template, changes to it would be non deterministic.

**Comment by anderbubble** [comment]
Good call about templates making this not really workable in most meaningful instances. Thanks for pointing that out.

**Comment by elfurbe** [comment]
I see it's now closed, but I feel like either I don't understand the reply or the purpose for what I intended must not have been clear, so I'll clarify into the void for posterity or whatever. Developing overlays is the purpose I have in mind, not viewing the rendered system state for a specific node. Being able to interactively drop into a chroot with an empty r/w overlayfs on top of an r/o base means you could interact with the system as necessary to install or edit or create or whatever you need to do in a normal interactive way, as if you were working on a system (or at least like you can work in any other chroot). When you exited the interactive session, all the changes to the filesystem would be contained within the r/w overlay. If you need to apply multiple overlays, the top-most overlay is r/w and all the others are r/o, that's how overlayfs works: https://www.kernel.org/doc/html/latest/filesystems/overlayfs.html?highlight=overlayfs#multiple-lower-layers 

I can see how rendering templated files out may be desirable for validating templated elements, so I don't think these ostensible features duplicate each other, though I can see how they would potentially be served by some shared architecture. I do not have an obvious answer for how one would mix template rendering that into what I'm describing, perhaps it's an intractable problem.

**Comment by anderbubble** [comment]
@elfurbe sorry about that; I had intended to reach out to you individually after closing this to discuss, and then I got pulled away into other tasks.

I understand that the goal is in support of overlay development. And, in theory, some of what you suggest would be feasible with some development; however, most overlays in Warewulf are the result of dynamically-rendered templates, not static files. As such, it seems that any attempt to provide an editing interface into overlays from the perspective of a fully-rendered image would be severely limited in capability, as edits to the resultant overlay renders could not then be backported into the templates that the overlay files were generated from.

On re-reading again, I see that you're talking here (at least in part) from a context of initializing a new overlay. That could be interesting; but any support like that would still necessarily only be useful for static files stored in an overlay; not for templated overlay files. I'm concerned that, at best, that significantly reduces the cost-benefit ratio of developing the feature; or at worst encourages incorrect patterns in the use of overlays. (e.g., rather than developing an overlay template, it might lead people down a path of manually developing many discrete per-node overlays)

It's possible I'm still missing something from what you're hoping to be able to do, so I'll reach out to you on Slack separately right after I send this. ^_^
